projects/netgrif-components-core/src/lib/data-fields/i18n-field/models/i18n-field.ts

```diff
@@ -4,15 +4,161 @@ import {Layout} from '../../models/layout';
 import {Validation} from '../../models/validation';
 import {Component} from '../../models/component';
 import {I18nFieldValue} from './i18n-field-value';
+import {Observable} from 'rxjs';
+import {FormControl, ValidatorFn} from '@angular/forms';
 
+export enum I18nFieldValidation {
+    TRANSLATION_REQUIRED = 'translationRequired',
+    TRANSLATION_ONLY = 'translationOnly',
+    REQUIRED_I18N = 'requiredI18n'
+}
 
 export class I18nField extends DataField<I18nFieldValue> {
 
+    public static toObject(templateValue: I18nFieldValue): any {
+        const object = {};
+        object['xx'] = templateValue.defaultValue;
+        for (const k in templateValue.translations) {
+            if (Object.prototype.hasOwnProperty.call(templateValue.translations, k)) {
+                object[k] = templateValue.translations[k];
+            }
+        }
+        return object;
+    }
+
+    public static fromObject(templateValue: any, templateKey: string): I18nFieldValue {
+        const i18nObject = {
+            defaultValue: templateValue['xx'],
+            key: templateKey,
+            translations: {}
+        };
+        for (const k in templateValue) {
+            if (k === 'xx') {
+                continue;
+            }
+            i18nObject.translations[k] = templateValue[k];
+        }
+        return i18nObject as I18nFieldValue;
+    }
+
     constructor(stringId: string, title: string, value: I18nFieldValue | string, behavior: Behavior, placeholder?: string,
                 description?: string, layout?: Layout, validations?: Array<Validation>, _component?: Component) {
         if (typeof value === 'string') {
             value = {defaultValue: value};
         }
         super(stringId, title, value, behavior, placeholder, description, layout, validations, _component);
     }
+
+    protected valueEquality(a: I18nFieldValue, b: I18nFieldValue): boolean {
+        if (!a && !b) {
+            return true;
+        }
+        if ((!a && !!b) || (!b && !!a)) {
+            return false;
+        }
+        if ((!a.defaultValue && !b.defaultValue) && (!a.translations && !b.translations) && (!a.key && !b.key)) {
+            return true;
+        }
+        if ((!(!a.defaultValue && !b.defaultValue)
```

Comment:
could this perhaps be made more readable?

Probably with calls to methods such as `hasTranslations()` and so on

Reply:
Done